### PR TITLE
[franceculture] improve extractor

### PR DESCRIPTION
**Pull request by flatgreen**
Hello,

This contribution enhance the 'FranceCulture' extractor.
Update test and extract a mp3 from the url site :
https://github.com/rg3/youtube-dl/issues/3777

thank you for your work :-)

### CODE REVIEW

File: youtube_dl/extractor/franceculture.py [hunk 1]
```diff
@@ -66,4 +71,43 @@ def _real_extract(self, url):
...
+    # extract '/player/reecouter?play=5093239' join to url base of franceculture
+    # extract mp3 with FranceCultureIE _extract_infos_from_player
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        video_path = self._html_search_regex(
+            r'<a class="rf-player-open".*?href="([^"]+)"', webpage, 'video path','no_path_player')
+        if video_path == 'no_path_player':
+            self.to_screen('no player : no sound in this page.')
```
Comment (jaimeMF) [inline youtube_dl/extractor/franceculture.py]: use instead
```pytho
raise ExtractorError('no player : no sound in this page.', expected=True)
```

File: youtube_dl/extractor/franceculture.py [hunk 2]
```diff
@@ -66,4 +71,43 @@ def _real_extract(self, url):
...
+            return None
+        new_id = re.search('play=(?P<id>[0-9]+)', video_path).group('id')
```
Comment (jaimeMF) [inline youtube_dl/extractor/franceculture.py]: Use `_search_regex`, it reports an error message if the regex doesn't match.

File: youtube_dl/extractor/franceculture.py [hunk 3]
```diff
@@ -66,4 +71,43 @@ def _real_extract(self, url):
...
+            'id': '5093239',
+            'display_id': 'les-carnets-de-la-creation-jean-gabriel-periot-cineaste-2015-10-13',
+            'ext': 'mp3',
+            'timestamp': 1444762500,
+            'upload_date': '20151013',
+            'description': 'startswith:Aujourd\'hui dans "Les carnets de la création", le cinéaste',
+        }
+    }
+
+    # dl url
```
Comment (jaimeMF) [inline youtube_dl/extractor/franceculture.py]: This comments are not needed, it's already clear from the code (you can add comment if there's something not obvious or if you have make some relevant choice)

File: youtube_dl/extractor/franceculture.py [hunk 4]
```diff
@@ -22,14 +24,14 @@ class FranceCultureIE(InfoExtractor):
...
-            'description': 'startswith:Avec :Jean-Baptiste Péretié pour son documentaire sur Arte "La revanche des « geeks », une enquête menée aux Etats',
+            'thumbnail': r're:^http://static\.franceculture\.fr/.*/images/player/Carnet-nomade\.jpg$',
+            'description': 'startswith:Avec :Jean-Baptiste Péretié pour son documentaire sur Arte "La revanche',
             'timestamp': 1393700400,
         }
     }
 
-    def _real_extract(self, url):
-        video_id = self._match_id(url)
+    def _extract_infos_from_player(self, url, video_id):
```
Comment (jaimeMF) [inline youtube_dl/extractor/franceculture.py]: Use `_extract_info_from_player`, without the `s`. Or just `_extract_info` or `_extract_from_player`, they are common in other extractors.

**Comment by jaimeMF** [comment]
About the extractor name, maybe it's better to use `FranceCultureEmission`?

**Comment by flatgreen** [comment]
Hello,

Here are the extractor 'franceculture' with the last changes.

Thank you

File: youtube_dl/extractor/franceculture.py [hunk 5]
```diff
@@ -66,4 +70,38 @@ def _real_extract(self, url):
             'alt_title': alt_title,
             'thumbnail': thumbnail,
             'description': description,
+            'display_id': display_id
         }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        return self._extract_infos_from_player(url, video_id)
```
Comment (jaimeMF) [inline youtube_dl/extractor/franceculture.py]: You forgot to change this.

File: youtube_dl/extractor/franceculture.py [hunk 6]
```diff
@@ -66,4 +70,38 @@ def _real_extract(self, url):
...
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        video_path = self._html_search_regex(
+            r'<a class="rf-player-open".*?href="([^"]+)"', webpage, 'video path', 'no_path_player')
+        if video_path == 'no_path_player':
+            raise ExtractorError('no player : no sound in this page.', expected=True)
+            return None
```
Comment (jaimeMF) [inline youtube_dl/extractor/franceculture.py]: The `return` is not needed anymore.

**Comment by flatgreen** [comment]
I hope all is well now, my tests seem good.

**Comment by jaimeMF** [comment]
Thanks, cherry picked in 7447661e9b433d8871d0a898c9af271d5a49ec82 and ecf17d165317078d0e31191e40903a74fc3da3e8.

Would you like to be credited in the AUTHORS file? If so, under which name?

**Comment by flatgreen** [comment]
You can write 'flatgreen' in the AUTHORS file.
Thank you for your patience, it was a pleasure :-)

**Pull request closed without merging**